File: src/presentation/playback.ts
/**
 * Timeline playback.
 *
 * Owns the clock. Advance it with the frame delta and ask what is active; the
 * renderer draws whatever comes back. Nothing here touches a canvas, so the
 * same scheduler drives a 2.5D renderer, a text log, or a test.
 *
 * Hitstop is handled here rather than in the renderer because it must freeze
 * the whole scene consistently - if each entity decided on its own, a hit
 * would smear.
 */

import type { AnimationStep, Timeline } from './timeline.js';
import { GLOBAL_TRACK } from './timeline.js';

export interface ActiveStep {
  step: AnimationStep;
  /** Normalized progress through this step, 0..1. */
  t: number;
  elapsedMs: number;
}

export type PlaybackState = 'idle' | 'playing' | 'finished';

export class Playback {
  private timeline: Timeline = { steps: [], durationMs: 0 };
  private clockMs = 0;
  /**
   * Clock value before the last `advance`. Starts below zero so that steps
   * scheduled at exactly time 0 are reported by the first `justStarted` call
   * rather than being skipped by an exclusive lower bound.
   */
  private previousClockMs = -1;
  /** Whether `advance` has run since the timeline was loaded. */
  private hasAdvanced = false;
  private hitstopMs = 0;
  private state: PlaybackState = 'idle';
  /** Playback rate. 2 is double speed, used by a "fast battle" option. */
  speed = 1;

  load(timeline: Timeline): void {
    this.timeline = timeline;
    this.clockMs = 0;
    this.previousClockMs = -1;
    this.hasAdvanced = false;
    this.hitstopMs = 0;
    this.state = timeline.steps.length > 0 ? 'playing' : 'finished';
  }

  /** Advance the clock. Returns the steps active after advancing. */
  advance(dtMs: number): ActiveStep[] {
    if (this.state !== 'playing') return [];

    let delta = Math.max(0, dtMs) * this.speed;
    // On the very first advance the previous bound stays below zero, so steps
    // scheduled at exactly time 0 still fall inside the (previous, current]
    // window. Afterwards it tracks the clock as it was before this frame.
    this.previousClockMs = this.hasAdvanced ? this.clockMs : -1;
    this.hasAdvanced = true;

    // Hitstop eats frame time before the clock sees it, freezing everything.
    if (this.hitstopMs > 0) {
      const consumed = Math.min(this.hitstopMs, delta);
      this.hitstopMs -= consumed;
      delta -= consumed;
      if (delta <= 0) return this.activeAt(this.clockMs);
    }

    this.clockMs += delta;
    if (this.clockMs >= this.timeline.durationMs) {
      this.clockMs = this.timeline.durationMs;
      this.state = 'finished';
    }
    return this.activeAt(this.clockMs);
  }

  /** Steps overlapping a given time. */
  activeAt(timeMs: number): ActiveStep[] {
    const active: ActiveStep[] = [];
    for (const step of this.timeline.steps) {
      if (timeMs < step.startMs) continue;
      const elapsedMs = timeMs - step.startMs;
      if (elapsedMs > step.durationMs) continue;
      const t = step.durationMs <= 0 ? 1 : Math.min(1, elapsedMs / step.durationMs);
      active.push({ step, t, elapsedMs });
    }
    return active;
  }

  /** Steps active for one entity, which is what a sprite needs each frame. */
  activeForTrack(track: string, timeMs = this.clockMs): ActiveStep[] {
    return this.activeAt(timeMs).filter((entry) => entry.step.track === track);
  }

  /**
   * Steps that began during the most recent `advance`, for one-shot triggers
   * like sound cues.
   *
   * The window is (previous clock, current clock], measured from the actual
   * last advance rather than a caller-supplied duration. That guarantees every
   * step fires exactly once: a caller-supplied window either drops cues when
   * it is shorter than the frame delta, or repeats them when it is longer.
   */
  justStarted(): AnimationStep[] {
    return this.timeline.steps.filter(
      (step) => step.startMs > this.previousClockMs && step.startMs <= this.clockMs,
    );
  }

  /** Request a scene-wide freeze. Called when a hitstop step begins. */
  requestHitstop(durationMs: number): void {
    this.hitstopMs = Math.max(this.hitstopMs, durationMs);
  }

  /** Jump to the end. For a "skip animation" button. */
  skip(): void {
    this.previousClockMs = this.hasAdvanced ? this.clockMs : -1;
    this.hasAdvanced = true;
    this.clockMs = this.timeline.durationMs;
    this.hitstopMs = 0;
    this.state = 'finished';
  }

  get isFinished(): boolean {
    return this.state === 'finished';
  }

  get isPlaying(): boolean {
    return this.state === 'playing';
  }

  get now(): number {
    return this.clockMs;
  }

  get totalMs(): number {
    return this.timeline.durationMs;
  }

  /** 0..1 through the whole timeline, for a progress indicator. */
  get progress(): number {
    if (this.timeline.durationMs <= 0) return 1;
    return this.clockMs / this.timeline.durationMs;
  }
}

/**
 * Collapse the steps active for one entity into a render state.
 *
 * Several steps can target the same entity at once - a flash and a shake and a
 * pose - so this folds them into the single set of values a sprite needs. Last
 * writer wins for discrete values like pose; continuous values accumulate.
 */
export interface EntityRenderState {
  pose: string;
  flashIntensity: number;
  flashColor: string;
  shakeOffset: number;
  opacity: number;
  lungeToward: string | null;
  lungeAmount: number;
  numerals: { text: string; style: string; t: number }[];
  bars: { stat: 'hp' | 'aether'; value: number; max: number }[];
  vfx: { effect: string; element: string | null; t: number }[];
}

export function emptyRenderState(): EntityRenderState {
  return {
    pose: 'idle',
    flashIntensity: 0,
    flashColor: '#ffffff',
    shakeOffset: 0,
    opacity: 1,
    lungeToward: null,
    lungeAmount: 0,
    numerals: [],
    bars: [],
    vfx: [],
  };
}

export function foldRenderState(active: readonly ActiveStep[]): EntityRenderState {
  const out = emptyRenderState();

  for (const { step, t } of active) {
    const action = step.action;
    switch (action.kind) {
      case 'pose':
        out.pose = action.pose;
        break;
      case 'flash':
        // Flash decays across its own duration.
        out.flashIntensity = Math.max(out.flashIntensity, action.intensity * (1 - t));
        out.flashColor = action.color;
        break;
      case 'shake': {
        // Decaying oscillation, so the shake settles rather than cutting out.
        const decay = 1 - t;
        out.shakeOffset += Math.sin(t * Math.PI * 8) * action.intensity * decay;
        break;
      }
      case 'lunge':
        out.lungeToward = action.towardId;
        // Out and back within the step.
        out.lungeAmount = Math.sin(t * Math.PI);
        break;
      case 'collapse':
        out.opacity = Math.min(out.opacity, 1 - t * 0.65);
        out.pose = 'down';
        break;
      case 'rise':
        out.opacity = Math.max(out.opacity, t);
        break;
      case 'numeral':
        out.numerals.push({ text: action.text, style: action.style, t });
        break;
      case 'bar': {
        const value = action.from + (action.to - action.from) * easeBar(t);
        out.bars.push({ stat: action.stat, value, max: action.max });
        break;
      }
      case 'vfx':
        out.vfx.push({ effect: action.effect, element: action.element, t });
        break;
      default:
        break;
    }
  }
  return out;
}

/** Bars drain fast then settle, which reads better than a linear slide. */
function easeBar(t: number): number {
  return 1 - Math.pow(1 - t, 3);
}

/** Scene-wide state folded from the global track. */
export interface SceneRenderState {
  banner: { text: string; tone: string; t: number } | null;
  screenFlash: number;
  cameraFocusId: string | null;
  hitstopActive: boolean;
}

export function foldSceneState(active: readonly ActiveStep[]): SceneRenderState {
  const out: SceneRenderState = { banner: null, screenFlash: 0, cameraFocusId: null, hitstopActive: false };
  for (const { step, t } of active) {
    if (step.track !== GLOBAL_TRACK) continue;
    const action = step.action;
    switch (action.kind) {
      case 'banner':
        out.banner = { text: action.text, tone: action.tone, t };
        break;
      case 'screen-flash':
        out.screenFlash = Math.max(out.screenFlash, 1 - t);
        break;
      case 'camera':
        out.cameraFocusId = action.focusId;
        break;
      case 'hitstop':
        out.hitstopActive = true;
        break;
      default:
        break;
    }
  }
  return out;
}

File: unity/Aetherlight.Core/Presentation/Playback.cs
using System;
using System.Collections.Generic;

namespace Aetherlight.Presentation
{
    public enum PlaybackState { Idle, Playing, Finished }

    public sealed class ActiveStep
    {
        public AnimationStep Step = new AnimationStep();
        /// <summary>Normalized progress through this step, 0..1.</summary>
        public double T;
        public double ElapsedMs;
    }

    /// <summary>
    /// Timeline playback: owns the clock.
    ///
    /// Advance it with the frame delta and ask what is active; the renderer
    /// draws whatever comes back. Nothing here touches an engine type, so the
    /// same scheduler drives a Unity renderer, a text log, or a test.
    ///
    /// Hitstop is handled here rather than in the renderer because it must
    /// freeze the whole scene consistently - if each entity decided on its own,
    /// a hit would smear.
    /// </summary>
    public sealed class Playback
    {
        private Timeline _timeline = new Timeline();
        private double _clockMs;

        /// <summary>
        /// Clock value before the last Advance. Starts below zero so steps
        /// scheduled at exactly time 0 are reported by the first JustStarted
        /// call rather than skipped by an exclusive lower bound.
        /// </summary>
        private double _previousClockMs = -1;

        /// <summary>Whether Advance has run since the timeline was loaded.</summary>
        private bool _hasAdvanced;

        private double _hitstopMs;
        private PlaybackState _state = PlaybackState.Idle;

        /// <summary>Playback rate. 2 is double speed, for a "fast battle" option.</summary>
        public double Speed { get; set; } = 1;

        public void Load(Timeline timeline)
        {
            _timeline = timeline;
            _clockMs = 0;
            _previousClockMs = -1;
            _hasAdvanced = false;
            _hitstopMs = 0;
            _state = timeline.Steps.Count > 0 ? PlaybackState.Playing : PlaybackState.Finished;
        }

        /// <summary>Advance the clock. Returns the steps active after advancing.</summary>
        public List<ActiveStep> Advance(double dtMs)
        {
            if (_state != PlaybackState.Playing) return new List<ActiveStep>();

            double delta = Math.Max(0, dtMs) * Speed;
            // On the very first advance the previous bound stays below zero, so
            // steps at exactly time 0 fall inside the (previous, current] window.
            _previousClockMs = _hasAdvanced ? _clockMs : -1;
            _hasAdvanced = true;

            // Hitstop eats frame time before the clock sees it, freezing everything.
            if (_hitstopMs > 0)
            {
                double consumed = Math.Min(_hitstopMs, delta);
                _hitstopMs -= consumed;
                delta -= consumed;
                if (delta <= 0) return ActiveAt(_clockMs);
            }

            _clockMs += delta;
            if (_clockMs >= _timeline.DurationMs)
            {
                _clockMs = _timeline.DurationMs;
                _state = PlaybackState.Finished;
            }
            return ActiveAt(_clockMs);
        }

        /// <summary>Steps overlapping a given time.</summary>
        public List<ActiveStep> ActiveAt(double timeMs)
        {
            var active = new List<ActiveStep>();
            foreach (var step in _timeline.Steps)
            {
                if (timeMs < step.StartMs) continue;
                double elapsed = timeMs - step.StartMs;
                if (elapsed > step.DurationMs) continue;
                double t = step.DurationMs <= 0 ? 1 : Math.Min(1, elapsed / step.DurationMs);
                active.Add(new ActiveStep { Step = step, T = t, ElapsedMs = elapsed });
            }
            return active;
        }

        /// <summary>Steps active for one entity, which is what a sprite needs each frame.</summary>
        public List<ActiveStep> ActiveForTrack(string track) => ActiveForTrack(track, _clockMs);

        public List<ActiveStep> ActiveForTrack(string track, double timeMs) =>
            ActiveAt(timeMs).FindAll(entry => entry.Step.Track == track);

        /// <summary>
        /// Steps that began during the most recent Advance, for one-shot
        /// triggers like sound cues.
        ///
        /// The window is (previous clock, current clock], measured from the
        /// actual last advance rather than a caller-supplied duration. That
        /// guarantees every step fires exactly once: a caller-supplied window
        /// either drops cues when shorter than the frame delta, or repeats them
        /// when longer.
        /// </summary>
        public List<AnimationStep> JustStarted() =>
            _timeline.Steps.FindAll(step => step.StartMs > _previousClockMs && step.StartMs <= _clockMs);

        /// <summary>Request a scene-wide freeze. Called when a hitstop step begins.</summary>
        public void RequestHitstop(double durationMs) => _hitstopMs = Math.Max(_hitstopMs, durationMs);

        /// <summary>Jump to the end, for a "skip animation" button.</summary>
        public void Skip()
        {
            _previousClockMs = _hasAdvanced ? _clockMs : -1;
            _hasAdvanced = true;
            _clockMs = _timeline.DurationMs;
            _hitstopMs = 0;
            _state = PlaybackState.Finished;
        }

        public bool IsFinished => _state == PlaybackState.Finished;
        public bool IsPlaying => _state == PlaybackState.Playing;
        public double Now => _clockMs;
        public double TotalMs => _timeline.DurationMs;

        /// <summary>0..1 through the whole timeline, for a progress indicator.</summary>
        public double Progress => _timeline.DurationMs <= 0 ? 1 : _clockMs / _timeline.DurationMs;
    }

    public sealed class NumeralView
    {
        public string Text = "";
        public NumeralStyle Style;
        public double T;
    }

    public sealed class BarView
    {
        public string Stat = "hp";
        public double Value;
        public double Max;
    }

    public sealed class VfxView
    {
        public string Effect = "";
        public Domain.Element? Element;
        public double T;
    }

    /// <summary>
    /// The values a sprite needs each frame, folded from every step currently
    /// targeting it. Several steps can overlap - a flash and a shake and a pose
    /// - so discrete values take the last writer and continuous ones accumulate.
    /// </summary>
    public sealed class EntityRenderState
    {
        public string Pose = "idle";
        public double FlashIntensity;
        public string FlashColor = "#ffffff";
        public double ShakeOffset;
        public double Opacity = 1;
        public string? LungeToward;
        public double LungeAmount;
        public List<NumeralView> Numerals = new List<NumeralView>();
        public List<BarView> Bars = new List<BarView>();
        public List<VfxView> Vfx = new List<VfxView>();
    }

    public sealed class SceneRenderState
    {
        public string? BannerText;
        public BannerTone BannerTone;
        public double BannerT;
        public bool HasBanner;
        public double ScreenFlash;
        public string? CameraFocusId;
        public bool HitstopActive;
    }

    public static class RenderFold
    {
        public static EntityRenderState ForEntity(IReadOnlyList<ActiveStep> active)
        {
            var outState = new EntityRenderState();

            foreach (var entry in active)
            {
                double t = entry.T;
                switch (entry.Step.Action)
                {
                    case PoseAction pose:
                        outState.Pose = pose.Pose;
                        break;

                    case FlashAction flash:
                        // Flash decays across its own duration.
                        outState.FlashIntensity = Math.Max(outState.FlashIntensity, flash.Intensity * (1 - t));
                        outState.FlashColor = flash.Color;
                        break;

                    case ShakeAction shake:
                        // Decaying oscillation, so the shake settles rather than cutting out.
                        outState.ShakeOffset += Math.Sin(t * Math.PI * 8) * shake.Intensity * (1 - t);
                        break;

                    case LungeAction lunge:
                        outState.LungeToward = lunge.TowardId;
                        outState.LungeAmount = Math.Sin(t * Math.PI);
                        break;

                    case CollapseAction:
                        outState.Opacity = Math.Min(outState.Opacity, 1 - t * 0.65);
                        outState.Pose = "down";
                        break;

                    case RiseAction:
                        outState.Opacity = Math.Max(outState.Opacity, t);
                        break;

                    case NumeralAction numeral:
                        outState.Numerals.Add(new NumeralView { Text = numeral.Text, Style = numeral.Style, T = t });
                        break;

                    case BarAction bar:
                        outState.Bars.Add(new BarView
                        {
                            Stat = bar.Stat,
                            Value = bar.From + (bar.To - bar.From) * EaseBar(t),
                            Max = bar.Max,
                        });
                        break;

                    case VfxAction vfx:
                        outState.Vfx.Add(new VfxView { Effect = vfx.Effect, Element = vfx.Element, T = t });
                        break;
                }
            }
            return outState;
        }

        /// <summary>Bars drain fast then settle, which reads better than a linear slide.</summary>
        private static double EaseBar(double t) => 1 - Math.Pow(1 - t, 3);

        public static SceneRenderState ForScene(IReadOnlyList<ActiveStep> active)
        {
            var outState = new SceneRenderState();
            foreach (var entry in active)
            {
                if (entry.Step.Track != TimelineBuilder.GlobalTrack) continue;
                switch (entry.Step.Action)
                {
                    case BannerAction banner:
                        outState.HasBanner = true;
                        outState.BannerText = banner.Text;
                        outState.BannerTone = banner.Tone;
                        outState.BannerT = entry.T;
                        break;
                    case ScreenFlashAction:
                        outState.ScreenFlash = Math.Max(outState.ScreenFlash, 1 - entry.T);
                        break;
                    case CameraAction camera:
                        outState.CameraFocusId = camera.FocusId;
                        break;
                    case HitstopAction:
                        outState.HitstopActive = true;
                        break;
                }
            }
            return outState;
        }
    }
}
